File: app/services/live_updates.py
"""Something-changed-go-check push notifications over Redis pub/sub —
what makes a honeypot's Overview/Monitoring/Updates tabs (and, more
broadly, the Dashboard, Map, Audit log, Companies list, and a user's own
Notification history) update the moment something relevant happens,
instead of waiting out their htmx polling interval (see
`app/web/routes/live_ws.py`, the WebSocket that relays these to the
browser, and `app/web/static/js/live-updates.js`, the client that turns
one into an htmx re-fetch of the matching panel).

Deliberately minimal: a published message carries only a `kind` — never
honeypot/audit/notification data. A client that receives one just
re-triggers the matching htmx panel's own `hx-get` (still permission/
scope-checked exactly like its periodic poll already was); every channel
here is a doorbell, not a data feed, so there's nothing sensitive in any
of them and no risk of a push payload drifting out of sync with what a
fresh fetch would show.

Four channel shapes, all sharing the same best-effort publish/relay
machinery:
- **Per-honeypot** (`channel_for`/`publish_honeypot_event`) — the
  original one, for a single honeypot's own tabs.
- **Fleet-wide** (`FLEET_CHANNEL`/`publish_fleet_event`) — "some
  honeypot's reachability or activity changed", fired alongside the
  per-honeypot publish for `KIND_STATUS`/`KIND_ACTIVITY` specifically
  (the only two kinds a cross-honeypot view like the Dashboard or Map
  actually cares about). Every logged-in user may subscribe — the
  payload carries no company/honeypot identity, so there's nothing to
  scope; the panel it triggers a re-fetch of is scoped on its own.
- **Admin-wide** (`ADMIN_CHANNEL`/`publish_admin_event`) — "a new audit
  log entry was written", fired from `app.audit.log_event` itself, so
  it also covers company/user create-edit-delete (always audit-logged).
  Superadmin-only subscription, matching the Audit log/Companies/Users
  pages this feeds.
- **Per-user notifications** (`notifications_channel_for`/
  `publish_notifications_event`) — "a NotificationLog row was just
  written for this user", fired from `app.services.notifications._log`.

Called from two different worlds: the FastAPI process (which has a
long-lived `app.state.redis`) and Celery worker processes (separate OS
processes entirely, each running these jobs' `asyncio.run(...)` wrapper —
see `app.tasks.jobs`'s module docstring). Rather than plumb `app.state`
through to worker code that has no `app` object, this opens its own
short-lived Redis connection per publish. That's more round trips than
reusing a shared connection, but publishes happen at most a few times per
job (not per SSH byte), and it keeps this module usable from anywhere
without a FastAPI request/app in scope.
"""

from __future__ import annotations

import json
import logging

import redis.asyncio as aioredis

from app.core.config import get_settings

logger = logging.getLogger(__name__)

_CHANNEL_PREFIX = "debcontrol:live:honeypot:"
FLEET_CHANNEL = "debcontrol:live:fleet"
ADMIN_CHANNEL = "debcontrol:live:admin"
_NOTIFICATIONS_CHANNEL_PREFIX = "debcontrol:live:notifications:"

# The finite set of "something changed" hints a client understands — see
# app/web/static/js/live-updates.js. Keeping this closed (rather than any
# free-form string) means a typo here fails loudly in review, not silently
# as a push nothing on the client ever matches.
Kind = str
KIND_STATUS = "status"
KIND_FACTS = "facts"
KIND_PACKAGES = "packages"
KIND_SERVICES = "services"
KIND_UPDATES = "updates"
KIND_MONITORING = "monitoring"
KIND_ACTIVITY = "activity"
KIND_AUDIT = "audit"
KIND_NOTIFICATION = "notification"


def channel_for(honeypot_id: str) -> str:
    return f"{_CHANNEL_PREFIX}{honeypot_id}"


def notifications_channel_for(user_id: str) -> str:
    return f"{_NOTIFICATIONS_CHANNEL_PREFIX}{user_id}"


async def _publish(channel: str, kind: Kind) -> None:
    """Best-effort — a failed publish just means the affected panel waits
    for its own polling fallback to catch up a little later. Never worth
    failing (or even slowing down) the actual job/request over, so every
    error is caught and logged, not raised."""
    try:
        client: aioredis.Redis = aioredis.from_url(  # type: ignore[no-untyped-call]
            get_settings().redis_url
        )
        try:
            await client.publish(channel, json.dumps({"kind": kind}))
        finally:
            await client.aclose()
    except Exception:  # noqa: BLE001 - best-effort, see docstring
        logger.warning("live_updates: failed to publish %r on %s", kind, channel)


async def publish_honeypot_event(honeypot_id: str, kind: Kind) -> None:
    await _publish(channel_for(honeypot_id), kind)


async def publish_fleet_event(kind: Kind) -> None:
    await _publish(FLEET_CHANNEL, kind)


async def publish_admin_event(kind: Kind = KIND_AUDIT) -> None:
    await _publish(ADMIN_CHANNEL, kind)


async def publish_notifications_event(user_id: str, kind: Kind = KIND_NOTIFICATION) -> None:
    await _publish(notifications_channel_for(user_id), kind)

File: app/web/static/js/live-updates.js
// "Something changed, go check" — the browser half of app/services/
// live_updates.py and app/web/routes/live_ws.py. Opens one reconnecting
// WebSocket per live-scoped anchor found on the page and turns each
// `{"kind": "..."}` message it receives into a plain DOM event
// (`live-<kind>`) dispatched on `document.body`. Every htmx panel that
// used to poll on a fixed interval now also listens for its matching
// event (`hx-trigger="every 60s, live-facts from:body"`, say) — the
// interval stays only as a fallback for a missed/dropped push, so panels
// update within roughly a second of a background job finishing instead
// of waiting out the old ~20s poll.
//
// Four independent anchors, each opening its own socket to its own route
// (a page sets at most one of these):
//   - `[data-live-honeypot-id]`  -> /honeypots/{id}/live/ws   (one honeypot's
//     Overview/Monitoring/Activity/Updates tabs — also offers the
//     backgrounded-tab browser-notification toggle, see below)
//   - `[data-live-fleet]`        -> /live/fleet/ws            (Dashboard, Map)
//   - `[data-live-admin]`        -> /live/admin/ws            (Audit log,
//     Companies list — superadmin only, same as those pages)
//   - `[data-live-notifications]`-> /live/notifications/ws    (a user's own
//     Notification history)
//
// No-ops entirely on a page with none of these anchors — nothing loads
// this unconditionally, each page opts in by including it.
//
// **Found live, previously undetected**: this whole script silently never
// worked — a debcontrol leftover looked for `[data-live-machine-id]` (this
// app's templates all set `data-live-honeypot-id` instead) and built the
// socket URL as `/machines/{id}/live/ws` (the real route, still, is
// `/honeypots/{id}/live/ws` — see app/web/routes/live_ws.py). The selector
// mismatch meant `anchor` was always null, so the whole file no-opped on
// every single page, on every load, since this app's first commit — no
// page's htmx panel ever received a push, everything ran on its polling
// fallback alone the whole time. Same class of bug as the `machine_ids`/
// `honeypot_ids` bulk-select mixup CLAUDE.md already documents elsewhere.
(() => {
  "use strict";

  const INITIAL_RETRY_MS = 1000;
  const MAX_RETRY_MS = 30000;

  // Opens `url`, dispatching `live-<kind>` on document.body for every
  // `{"kind": "..."}` text frame received, and reconnects with capped
  // exponential backoff on any close/error — forever, since these sockets
  // are meant to outlive the page. `onKind(kind)` is called for the same
  // message, for callers that need the raw kind too (the honeypot-scoped
  // connection's background-notification logic below); other callers pass
  // a no-op.
  function openLiveSocket(url, onKind) {
    let retryDelayMs = INITIAL_RETRY_MS;
    let retryTimer = null;

    function scheduleReconnect() {
      if (retryTimer !== null) return; // already scheduled
      retryTimer = window.setTimeout(() => {
        retryTimer = null;
        connect();
      }, retryDelayMs);
      retryDelayMs = Math.min(retryDelayMs * 2, MAX_RETRY_MS);
    }

    function handleMessage(event) {
      let payload;
      try {
        payload = JSON.parse(event.data);
      } catch {
        return; // not JSON — ignore rather than crash the socket handler
      }
      if (!payload || typeof payload.kind !== "string") return;
      document.body.dispatchEvent(new Event(`live-${payload.kind}`));
      onKind(payload.kind);
    }

    function connect() {
      let socket;
      try {
        socket = new WebSocket(url);
      } catch {
        scheduleReconnect();
        return;
      }
      socket.addEventListener("open", () => {
        retryDelayMs = INITIAL_RETRY_MS; // a successful connection resets backoff
      });
      socket.addEventListener("message", handleMessage);
      socket.addEventListener("close", scheduleReconnect);
      // A socket that errors also fires "close" right after — closing it
      // explicitly here just avoids waiting on the browser's own timeout.
      socket.addEventListener("error", () => socket.close());
    }

    connect();

    // A backgrounded tab's WebSocket can go stale (some browsers/proxies
    // silently drop long-idle connections without ever firing "close") —
    // reconnecting on visibility is cheap insurance, not a fix for anything
    // observed, and the server-side idle cap is the real backstop either way.
    document.addEventListener("visibilitychange", () => {
      if (document.visibilityState === "visible") {
        retryDelayMs = INITIAL_RETRY_MS;
      }
    });
  }

  const scheme = window.location.protocol === "https:" ? "wss" : "ws";
  const host = window.location.host;

  // --- Honeypot-scoped: Overview/Monitoring/Activity/Updates tabs -------

  const honeypotAnchor = document.querySelector("[data-live-honeypot-id]");
  if (honeypotAnchor) {
    const honeypotId = honeypotAnchor.getAttribute("data-live-honeypot-id");
    const honeypotName =
      honeypotAnchor.getAttribute("data-live-honeypot-name") || "This honeypot";
    const url = `${scheme}://${host}/honeypots/${encodeURIComponent(honeypotId)}/live/ws`;

    // --- Browser notifications, for the same event when this tab is
    // backgrounded (`document.visibilityState === "hidden"`) and the
    // viewer opted in via the toggle button this injects next to the page
    // heading. Deliberately the plain Notification API, not the Push API:
    // no service worker, no server-side subscription storage, nothing
    // that would still fire with the tab fully closed. It only ever
    // surfaces something this same open tab already received over the
    // WebSocket above.

    const NOTIFY_PREF_KEY = "honeypotshelf:notifications-enabled";

    const KIND_MESSAGES = {
      status: "Reachability status changed",
      facts: "Facts refreshed",
      packages: "Installed packages refreshed",
      services: "Services refreshed",
      updates: "Update availability changed",
      monitoring: "New monitoring sample",
      activity: "New OpenCanary activity",
    };

    function notificationsWanted() {
      try {
        return window.localStorage.getItem(NOTIFY_PREF_KEY) === "1";
      } catch {
        return false; // private browsing / storage blocked — just skip it
      }
    }

    function setNotificationsWanted(value) {
      try {
        window.localStorage.setItem(NOTIFY_PREF_KEY, value ? "1" : "0");
      } catch {
        // Nothing to persist to — the toggle still reflects the in-memory
        // choice for the rest of this page view, it just won't survive a
        // reload. Not worth surfacing an error for.
      }
    }

    function notifyIfBackgrounded(kind) {
      if (!("Notification" in window)) return;
      if (Notification.permission !== "granted") return;
      if (!notificationsWanted()) return;
      if (document.visibilityState !== "hidden") return; // tab is frontmost — the DOM update is enough
      const body = KIND_MESSAGES[kind] || "Something changed";
      let notification;
      try {
        notification = new Notification(honeypotName, { body, tag: `honeypotshelf-${honeypotId}` });
      } catch {
        return; // some browsers throw if constructed from a background/service context
      }
      notification.onclick = () => {
        window.focus();
        notification.close();
      };
    }

    // --- The toggle button, injected next to the page heading rather than
    // duplicated in three separate templates. ---

    function buildToggle() {
      if (!("Notification" in window)) return; // unsupported browser — nothing to offer

      const button = document.createElement("button");
      button.type = "button";
      button.className = "link-button live-notify-toggle";

      function render() {
        if (Notification.permission === "denied") {
          button.textContent = "🔕 Notifications blocked";
          button.disabled = true;
          button.title = "Blocked in this browser's site settings.";
          return;
        }
        if (Notification.permission === "granted" && notificationsWanted()) {
          button.textContent = "🔔 Notifications on";
          button.title = "Click to turn off background notifications for this machine's page.";
        } else {
          button.textContent = "🔔 Enable notifications";
          button.title = "Get a browser notification when this page updates while backgrounded.";
        }
      }

      button.addEventListener("click", async () => {
        if (Notification.permission === "default") {
          let permission;
          try {
            permission = await Notification.requestPermission();
          } catch {
            return;
          }
          if (permission === "granted") setNotificationsWanted(true);
          render();
          return;
        }
        if (Notification.permission === "granted") {
          setNotificationsWanted(!notificationsWanted());
          render();
        }
      });

      render();
      honeypotAnchor.appendChild(button);
    }

    openLiveSocket(url, notifyIfBackgrounded);
    buildToggle();
  }

  // --- Fleet-wide: Dashboard, Map ----------------------------------------
  // "Some honeypot's reachability or activity changed somewhere in view" —
  // no per-connection scoping needed, see app/services/live_updates.py's
  // module docstring for why that's safe (the payload carries no identity;
  // the panel it re-fetches is scoped on its own).

  if (document.querySelector("[data-live-fleet]")) {
    openLiveSocket(`${scheme}://${host}/live/fleet/ws`, () => {});
  }

  // --- Admin-wide: Audit log, Companies list ------------------------------
  // "A new audit log entry was written" — covers company/user
  // create/edit/delete too, since those are always audit-logged.

  if (document.querySelector("[data-live-admin]")) {
    openLiveSocket(`${scheme}://${host}/live/admin/ws`, () => {});
  }

  // --- Per-user: Notification history -------------------------------------

  if (document.querySelector("[data-live-notifications]")) {
    openLiveSocket(`${scheme}://${host}/live/notifications/ws`, () => {});
  }
})();
